apollo-federation/src/sources/connect/json_selection/selection_set.rs

```diff
@@ -669,4 +691,53 @@ mod tests {
 }"###
         );
     }
+
+    #[test]
+    fn test_ensuring_key_fields() {
+        let json = super::JSONSelection::parse(
+            r###"
+            id
+            store { id }
+            name
+            price
+            "###,
+        )
+        .unwrap();
+
+        let schema = Schema::parse_and_validate(
+            r###"
+        type Query {
+            product: Product
+        }
+
+        type Product {
+            id: ID!
+            store: Store!
+            name: String!
+            price: String
+        }
+
+        type Store {
+          id: ID!
+        }
+        "###,
+            "./",
+        )
+        .unwrap();
+
+        let (document, selection_set) = selection_set(&schema, "{ product { name price } }");
+
+        let keys =
+            FieldSet::parse_and_validate(&schema, name!(Product), "id store { id }", "").unwrap();
```

Comment:
We still need a general way of extracting these key `FieldSet`s from a `JSONSelection`, right?

Reply:
yeah, this currently only works with something like `{$this.id}/{$this.store.id}` so we still need a way to support `$this { id store { id } }`